From 67eb032ce5df5a956d440dc55695739de5bc17d5 Mon Sep 17 00:00:00 2001
From: Kristian Amlie <kristian.amlie@northern.tech>
Date: Tue, 20 Feb 2018 13:52:07 +0100
Subject: [PATCH] vexpress-qemu specific fixes for Flash.

These are only needed because vexpress-qemu is not originally a Flash
based board.

Signed-off-by: Kristian Amlie <kristian.amlie@northern.tech>
---
 configs/vexpress_ca9x4_defconfig  |  5 ++++-
 include/configs/vexpress_ca9x4.h  | 10 ++++++++++
 include/configs/vexpress_common.h |  3 +--
 3 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/configs/vexpress_ca9x4_defconfig b/configs/vexpress_ca9x4_defconfig
index f10e73f86e..73767e719f 100644
--- a/configs/vexpress_ca9x4_defconfig
+++ b/configs/vexpress_ca9x4_defconfig
@@ -1,7 +1,7 @@
 CONFIG_ARM=y
 CONFIG_TARGET_VEXPRESS_CA9X4=y
 CONFIG_SYS_TEXT_BASE=0x60800000
-CONFIG_ENV_SIZE=0x40000
+CONFIG_ENV_SIZE=0x20000
 CONFIG_NR_DRAM_BANKS=2
 CONFIG_ENV_SECT_SIZE=0x40000
 CONFIG_DISTRO_DEFAULTS=y
@@ -24,6 +24,8 @@ CONFIG_ENV_IS_IN_FLASH=y
 CONFIG_ENV_ADDR=0x47F80000
 CONFIG_MTD=y
 CONFIG_MTD_NOR_FLASH=y
+CONFIG_MTDIDS_DEFAULT="nor2=40000000.flash"
+CONFIG_MTDPARTS_DEFAULT="mtdparts=40000000.flash:1m(u-boot)ro,-(ubi)"
 CONFIG_FLASH_CFI_DRIVER=y
 CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
 CONFIG_SYS_FLASH_PROTECTION=y
@@ -35,3 +36,5 @@ CONFIG_BAUDRATE=38400
 CONFIG_CONS_INDEX=0
 CONFIG_OF_LIBFDT=y
 # CONFIG_EFI_LOADER is not set
+CONFIG_RBTREE=y
+CONFIG_LZO=y
diff --git a/include/configs/vexpress_ca9x4.h b/include/configs/vexpress_ca9x4.h
index 8157a5868d..9c3ef14d14 100644
--- a/include/configs/vexpress_ca9x4.h
+++ b/include/configs/vexpress_ca9x4.h
@@ -13,4 +13,17 @@
 #define CONFIG_VEXPRESS_ORIGINAL_MEMORY_MAP
 #include "vexpress_common.h"
 
+#ifdef CONFIG_SYS_MALLOC_LEN
+#undef CONFIG_SYS_MALLOC_LEN
+#define CONFIG_SYS_MALLOC_LEN (8 * 1024 * 1024)
+#endif
+
+#define CONFIG_SYS_FLASH_BASE 0x40000000
+
+#define CONFIG_MTD_CONCAT
+#define CONFIG_FLASH_CFI_MTD
+#define CONFIG_BOOTCOUNT_LIMIT
+#define CONFIG_BOOTCOUNT_ENV
+#define CONFIG_SYS_REDUNDAND_ENVIRONMENT
+
 #endif /* VEXPRESS_CA9X4_H */
diff --git a/include/configs/vexpress_common.h b/include/configs/vexpress_common.h
index b124f00c9d..f5c13accfe 100644
--- a/include/configs/vexpress_common.h
+++ b/include/configs/vexpress_common.h
@@ -202,8 +202,7 @@
 		"console=ttyAMA0,38400n8\0" \
 		"dram=1024M\0" \
 		"root=/dev/sda1 rw\0" \
-		"mtd=armflash:1M@0x800000(uboot),7M@0x1000000(kernel)," \
-			"24M@0x2000000(initrd)\0" \
+		"mtdparts=" CONFIG_MTDPARTS_DEFAULT "\0" \
 		"flashargs=setenv bootargs root=${root} console=${console} " \
 			"mem=${dram} mtdparts=${mtd} mmci.fmax=190000 " \
 			"devtmpfs.mount=0  vmalloc=256M\0" \
-- 
2.17.1

